stdlib/public/SDK/Foundation/NSFastEnumeration.swift

```diff
@@ -16,63 +16,35 @@
 //===----------------------------------------------------------------------===//
 // Fast enumeration
 //===----------------------------------------------------------------------===//
-
-// NB: This is a class because fast enumeration passes around interior pointers
-// to the enumeration state, so the state cannot be moved in memory. We will
-// probably need to implement fast enumeration in the compiler as a primitive
-// to implement it both correctly and efficiently.
-final public class NSFastEnumerationIterator : IteratorProtocol {
-  var enumerable: NSFastEnumeration
-  var state: [NSFastEnumerationState]
-  var n: Int
-  var count: Int
-
-  /// Size of ObjectsBuffer, in ids.
-  static var STACK_BUF_SIZE: Int { return 4 }
-
-  var objects: [Unmanaged<AnyObject>?]
-
-  public func next() -> Any? {
-    if n == count {
-      // FIXME: Is this check necessary before refresh()?
-      if count == 0 { return nil }
-      refresh()
-      if count == 0 { return nil }
+public struct NSFastEnumerationIterator : IteratorProtocol {
+    var enumerable: NSFastEnumeration
+    var objects: (Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?) = (nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
+    var state = NSFastEnumerationState(state: 0, itemsPtr: nil, mutationsPtr: _fastEnumerationStorageMutationsPtr, extra: (0, 0, 0, 0, 0))
+    var index = 0
+    var count = 0
+    
+    public init(_ enumerable: NSFastEnumeration) {
+        self.enumerable = enumerable
     }
-    let next: Any = state[0].itemsPtr![n]!
-    n += 1
-    return next
-  }
-
-  func refresh() {
-    _sanityCheck(objects.count > 0)
-    n = 0
-    objects.withUnsafeMutableBufferPointer {
-      count = enumerable.countByEnumerating(
-        with: &state,
-        objects: AutoreleasingUnsafeMutablePointer($0.baseAddress!),
-        count: $0.count)
+    
+    public mutating func next() -> Any? {
+        if index + 1 > count {
+            index = 0
+            count = withUnsafeMutablePointer(to: &objects) {
+                return enumerable.countByEnumerating(with: &state, objects: AutoreleasingUnsafeMutablePointer($0), count: 16)
```

Comment:
err the two param version no longer exists?

I don't get the claim that it might not be safe, I was under the assumption the tombstone is dropped in the outer structure (the Iterator) for objects when & is used. Then in the inner block the tombstone for state is dropped in the outer structure (the iterator). The memory layout of the structure containing ought to not have changed when they are restored unless the structure is somehow dragged somewhere else in memory during the call (which would be a feat and a half).

My only concern was the case of the objects being unmanaged as declaration and converted into an AUMP. But that seems to be a non issue from what I can tell.

```swift
var deadCount = 0
let total = 100
class Tracker {
    deinit {
        deadCount += 1
    }
}

autoreleasepool {
    var iter = NSFastEnumerationIterator(NSArray(array: (0..<total).map { _ in Tracker() }))
    var finished = false
    repeat {
        autoreleasepool {
            for _ in 0..<20 {
                finished = iter.next() == nil
            }
        }
    } while !finished
}

assert(deadCount == total)
```